src/smtp_observer.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The SMTP integrator relation observer."""
+
+# ignoring duplicate-code with container connect check in the saml observer.
+# pylint: disable=R0801
+
+import logging
+import typing
+from typing import Optional
+
+import ops
+from charms.smtp_integrator.v0.smtp import (
+    AuthType,
+    SmtpDataAvailableEvent,
+    SmtpRelationData,
+    SmtpRequires,
+    TransportSecurity,
+)
+from ops.charm import CharmBase
+from ops.framework import Object
+from pydantic import ValidationError
+
+import synapse
+from charm_types import SMTPConfiguration
+from pebble import PebbleServiceError
+
+logger = logging.getLogger(__name__)
+
+
+class SMTPObserver(Object):
+    """The SMTP relation observer.
+
+    Attrs:
+        _pebble_service: instance of pebble service.
+    """
+
+    _RELATION_NAME = "smtp"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "smtp-observer")
+        self._charm = charm
+        self.smtp = SmtpRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+        )
+        self.framework.observe(
+            self.smtp.on.smtp_data_available,
+            self._on_smtp_relation_data_available,
+        )
+
+    def get_relation_as_smtp_conf(self) -> Optional[SMTPConfiguration]:
+        """Get SMTP data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+        """
+        try:
+            relation_data: Optional[SmtpRelationData] = self.smtp.get_relation_data()
+        except ValidationError:
+            # ValidationError happens in the smtp(_legacy)relation_created event, as
+            # the relation databag is empty at that point.
+            logger.info("SMTP databag is empty. SMTP information will be set in the next event.")
+            return None
+
+        if relation_data is None:
+            return None
+
+        user = None
+        password = None
+        if relation_data.auth_type == AuthType.PLAIN:
+            user = relation_data.user
+            password = self._get_password_from_relation_data(relation_data)
```

Comment:
This is where we should consider what kind of auth is ok. When we designed the SMTP integrator, we decided to be more flexible and let the users of the integration decide. In general, we should be careful with that we support with security since it reflects poorly on the charming ecosystem if within it it is common to have insecure deployments. Any thoughts @gregory-schiano?

Reply:
I'm fine not supporting plain auth in synapse charm as it supports STARTTLS and TLS which are more secure standards.
We could mark the status as Blocked with a message stating that we don't support plain auth